### ISSUE TITLE
Submitted report version should be immutable

### ISSUE THREAD

**Issue by pbastia** [bcgov/cas-reporting#459]
#### Description:

As a developer, I want the submitted report data to be protected by database triggers, so that accidental edits of submitted data is not possible.

Prevent modification of report data where the report version is submitted
- use: https://django-pgtrigger.readthedocs.io/en/4.13.3/
- The base class for report models is probably where we want to start
- Check with m2m through tables (See comment below)

Note: if starting this work before #416, the extension needs to be installed in a first small PR, so the work is not blocked

#### Acceptance Criteria:

Given I am a developer
When I edit a report version or any of its dependent models in the database
And that report version is 'submitted'
Then a trigger throws an error

#### Development Checklist:

- [ ] Install aforementioned django app
- [ ] Create a base trigger for the behaviour
- [ ] Associate it with each model where needed
- [ ] Test that every model part of a report version has the trigger on its `Meta` class. Note: there is a test that lists all cascading models from the report version, that piece can be re-used. For questions ask @pbastia 

#### Definition of Ready (Note: If any of these points are not applicable, mark N/A)

- [ ] User story is included
- [ ] User role and type are identified
- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if required)
- [ ] Design / Solution is accepted by Product Owner
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

#### Definition of Done (Note: If any of these points are not applicable, mark N/A)

- [ ] Acceptance criteria are tested by the CI pipeline
- [ ] UI meets accessibility requirements
- [ ] Configuration changes are documented, documentation and designs are updated
- [ ] Passes code peer-review
- [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
- [ ] Ticket is ready to be merged to main branch
- [ ] Can be demoed in Sprint Review
- [ ] Bugs or future work cards are identified and created
- [ ] Reviewed and approved by Product Owner

#### Notes:

-

**Comment by rdromey** [comment]
@ayeshmcg to write dev checklist and AC

**Comment by Sepehr-Sobhani** [comment]
We have three options:

#### Option 1: Overriding `save` and `update` in Django ORM  
Enforce immutability directly in Django by overriding the `save` method to block `.save()` operations and customizing the `update` method in a `QuerySet` to block `.update()` calls when `status='Submitted'`.

Example:

```python
class MyModelQuerySet(models.QuerySet):
    def update(self, **kwargs):
        if self.filter(status='Submitted').exists():
            raise ValidationError("Cannot update objects with status 'Submitted'.")
        return super().update(**kwargs)

class MyModelManager(models.Manager):
    def get_queryset(self):
        return MyModelQuerySet(self.model, using=self._db)

class MyModel(models.Model):
    status = models.CharField(max_length=20)
    objects = MyModelManager()

    def save(self, *args, **kwargs):
        if self.pk and MyModel.objects.get(pk=self.pk).status == 'Submitted':
            raise ValidationError("Cannot modify an object with status 'Submitted'.")
        super().save(*args, **kwargs)
```
**This approach works well within Django ORM but does not protect against raw SQL updates.**

#### Option 2: Using `django-pgtrigger`  
Define a PostgreSQL trigger declaratively in Python using [django-pgtrigger](https://django-pgtrigger.readthedocs.io/en/4.13.3/) to block updates for rows where `status='Submitted'`.  
Example:  
```python
    class Meta:
        triggers = [
            pgtrigger.Protect(
                name="protect_submitted_updates",
                operation=pgtrigger.Update,
                condition=pgtrigger.Q(old__status='Submitted'),
            )
        ]
```

**This enforces immutability at the database level, making it effective against raw SQL and external tools.** (NOT TESTED YET)

#### Option 3: Adding a Raw SQL Trigger in a Migration  
Use a migration to manually define a PostgreSQL trigger, ensuring immutability directly at the database level.  
Example:  
```sql
CREATE OR REPLACE FUNCTION prevent_update_if_submitted()
RETURNS TRIGGER AS $$
BEGIN
    IF OLD.status = 'Submitted' THEN
        RAISE EXCEPTION 'Cannot modify an object with status Submitted.';
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;
```
**This ensures robust enforcement but requires managing raw SQL logic.**


### Recommendation
My suggestion would be Option 1, as it closely aligns with Django’s ORM and keeps logic within the application layer. However, Options 2 and 3 are equally valid if database-level enforcement is desired.

**Comment by pbastia** [comment]
Thank you @Sepehr-Sobhani !

I agree with your recommendation, however, I think Option#2 would be better suited for our needs, I think we're interested in avoiding any path where we'd accidentally update a submitted report. 
There are multiple failure points we might run into, for example: 
- if another application connects to the data,
- if a developer accidentally manipulates database records when fixing unrelated issues, 
- if we accidentally use an ORM method we didn't override (`update_or_create`, `bulk_create` on dependent records, etc.)

I really like how django-pgtrigger re-uses the django orm paradigms to configure those triggers, and how self-contained that behaviour will be.

**Comment by pbastia** [comment]
To test:
- whether django-pgtrigger triggers the trigger on an m2m update (i.e. facility report associated products). If that's the case, then we don't need to cover this case yet. Otherwise, we need to find a way (either change the through table to a manual one, or have django let us access it to add a trigger somehow)

**Comment by pbastia** [comment]
Decision with @Sepehr-Sobhani and @dleard : Use django-pgtrigger

**Comment by Sepehr-Sobhani** [comment]
This ticket is unblocked by merging PR: https://github.com/bcgov/cas-registration/pull/2729
